### PR TITLE
base farm claim impl

### PR DESCRIPTION
**Pull request by dorin-iancu**

### CODE REVIEW

File: auto-farm/src/user_farm_tokens.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+                tokens.push(payment);
+            }
+        });
+    }
+
+    #[endpoint(withdrawFarmTokens)]
+    fn withdraw_farm_tokens(&self) -> PaymentsVec<Self::Api> {
+        let caller = self.blockchain().get_caller();
+        let user_id = self.user_ids().get_id_or_insert(&caller);
```
Comment (CostinCarabas) [inline auto-farm/src/user_farm_tokens.rs:34]: I think you shouldn't have used `get_id_or_insert` here. If there is no known id, the endpoint should fail.

File: auto-farm/src/user_farm_tokens.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        if !tokens.is_empty() {
+            self.send().direct_multi(&caller, &tokens);
+        }
+
+        tokens
+    }
+
+    #[view(getUserFarmTokens)]
+    fn get_user_farm_tokens_view(&self, user: ManagedAddress) -> PaymentsVec<Self::Api> {
+        let user_id = self.user_ids().get_id_or_insert(&user);
```
Comment (CostinCarabas) [inline auto-farm/src/user_farm_tokens.rs:45]: Same here. You shouldn't have used `get_id_or_insert`.

File: auto-farm/src/user_farm_actions.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    + crate::user_farm_tokens::UserFarmTokensModule
+    + utils::UtilsModule
+{
+    /// Arg: user to claim rewards for
+    #[endpoint(claimAllFarmRewards)]
+    fn claim_all_farm_rewards(&self, user: ManagedAddress) {
+        self.require_caller_proxy_claim_address();
+
+        let farms_mapper = self.farm_ids();
+        let user_id = self.user_ids().get_id_or_insert(&user);
```
Comment (CostinCarabas) [inline auto-farm/src/user_farm_actions.rs:22]: If the user is not in the `user_ids` you shouldn't `insert` the user; the endpoint should fail.

File: auto-farm/src/user_farm_actions.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        let user_tokens_mapper = self.user_farm_tokens(user_id);
+
+        let user_tokens = user_tokens_mapper.get();
+        let mut new_user_farm_tokens = PaymentsVec::new();
```
Comment (CostinCarabas) [inline auto-farm/src/user_farm_actions.rs:26]: What is `new_user_farm_tokens` used for?

File: auto-farm/src/farms_whitelist.rs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            self.require_sc_address(&farm_addr);
+
+            let new_id = farms_mapper.get_id_or_insert(&farm_addr);
+            if new_id == NULL_ID {
+                continue;
+            }
+
+            let farm_config = self.get_farm_config(&farm_addr);
+            self.farm_for_farm_token(&farm_config.farm_token_id)
+                .set(new_id);
```
Comment (CostinCarabas) [inline auto-farm/src/farms_whitelist.rs:23]: If `new_id` already exists, wouldn't `set` fail?
`get_id_or_insert` returns an existing id -> set; else -> update

Comment (CostinCarabas) [reply auto-farm/src/farms_whitelist.rs:23]: Also for `.insert` below.

**Review by CostinCarabas — CHANGES_REQUESTED** [review]

File: auto-farm/src/user_farm_tokens.rs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    #[endpoint(depositFarmTokens)]
+    fn deposit_farm_tokens(&self) {
+        let payments = self.get_non_empty_payments();
+        let caller = self.blockchain().get_caller();
+        let user_id = self.user_ids().get_id_or_insert(&caller);
+
+        self.user_farm_tokens(user_id).update(|tokens| {
+            for payment in &payments {
+                let farm_id = self.farm_for_farm_token(&payment.token_identifier).get();
+                require!(farm_id != NULL_ID, "Invalid token");
```
[the remark this answers is not in the document]
Comment (dorin-iancu) [reply auto-farm/src/user_farm_tokens.rs:24]: `user_farm_tokens` is only updated from payments, so we're guaranteed by the protocol to have valid token IDs.

File: auto-farm/src/user_farm_actions.rs (hunk 4 above)
Comment (dorin-iancu) [reply auto-farm/src/user_farm_actions.rs:26]: After every farm action, user receives a new farm token, so we need to save an updated list with the new tokens. Additionally, farm tokens that have no associated farm address (which can happen if a farm is removed or paused) will be left unchanged.

File: auto-farm/src/user_farm_tokens.rs (hunk 1 above)
Comment (dorin-iancu) [reply auto-farm/src/user_farm_tokens.rs]: Fixed.

File: auto-farm/src/user_farm_tokens.rs (hunk 2 above)
Comment (dorin-iancu) [reply auto-farm/src/user_farm_tokens.rs]: Fixed.

File: auto-farm/src/user_farm_actions.rs (hunk 3 above)
Comment (dorin-iancu) [reply auto-farm/src/user_farm_actions.rs]: Fixed.

File: auto-farm/src/farms_whitelist.rs (hunk 5 above)
Comment (dorin-iancu) [reply auto-farm/src/farms_whitelist.rs:24]: Fixed, it would always add farm addresses even if they were existing already.

**Review by psorinionut — APPROVED** [review]

**Review by CostinCarabas — APPROVED** [review]

**Pull request merged**